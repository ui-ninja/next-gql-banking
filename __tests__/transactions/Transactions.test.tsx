import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom';
import Transactions from '../../pages/transactions/[accountId]';

const dummyTransactions = {
  transactions: {
    edges: [
      {
        cursor: '63b84ad9fc13ae2b950000b4',
        node: {
          id: '63b84ad9fc13ae2b950000b4',
          type: 'credit',
          createdAt: '2021-02-21T00:05:34Z',
          amount: 80894.47,
          accountId: '0af25fe0-48a7-4bfe-a556-8a793255ef70',
          rowNo: 1,
        },
      },
      {
        cursor: '63b84ad9fc13ae2b950000b5',
        node: {
          id: '63b84ad9fc13ae2b950000b5',
          type: 'credit',
          createdAt: '2019-04-21T09:09:46Z',
          amount: 39227.57,
          accountId: '4ec8c52c-2a36-4bf2-9021-14697a439f60',
          rowNo: 2,
        },
      },
      {
        cursor: '63b84ad9fc13ae2b950000b6',
        node: {
          id: '63b84ad9fc13ae2b950000b6',
          type: 'credit',
          createdAt: '2021-02-06T13:04:12Z',
          amount: 86036.65,
          accountId: '8399c3e2-d73f-4478-99ef-3ac7f7aa1380',
          rowNo: 3,
        },
      },
      {
        cursor: '63b84ad9fc13ae2b950000b7',
        node: {
          id: '63b84ad9fc13ae2b950000b7',
          type: 'credit',
          createdAt: '2021-02-15T05:17:32Z',
          amount: 7464.81,
          accountId: 'dc87cdd5-d8c8-4337-9269-027b1ffcb77a',
          rowNo: 4,
        },
      },
      {
        cursor: '63b84ad9fc13ae2b950000b8',
        node: {
          id: '63b84ad9fc13ae2b950000b8',
          type: 'credit',
          createdAt: '2019-09-06T06:20:27Z',
          amount: 7113.16,
          accountId: 'f9669e5e-92a2-49b0-beeb-37915b32c7a8',
          rowNo: 5,
        },
      },
      {
        cursor: '63b84ad9fc13ae2b950000b9',
        node: {
          id: '63b84ad9fc13ae2b950000b9',
          type: 'credit',
          createdAt: '2019-10-09T17:12:04Z',
          amount: 12840.67,
          accountId: '2a35b454-ab9f-4bba-bb52-d38d8d24cff0',
          rowNo: 6,
        },
      },
      {
        cursor: '63b84ad9fc13ae2b950000ba',
        node: {
          id: '63b84ad9fc13ae2b950000ba',
          type: 'debit',
          createdAt: '2023-08-24T02:19:50Z',
          amount: 18654.31,
          accountId: 'f66cdf81-1d87-4529-9916-2d17bdacd859',
          rowNo: 7,
        },
      },
      {
        cursor: '63b84ad9fc13ae2b950000bb',
        node: {
          id: '63b84ad9fc13ae2b950000bb',
          type: 'debit',
          createdAt: '2023-07-17T18:31:06Z',
          amount: 14473.13,
          accountId: '6931cd1a-8514-485d-b1ec-44760c02a511',
          rowNo: 8,
        },
      },
      {
        cursor: '63b84ad9fc13ae2b950000bc',
        node: {
          id: '63b84ad9fc13ae2b950000bc',
          type: 'debit',
          createdAt: '2019-01-24T12:20:04Z',
          amount: 96879.76,
          accountId: '1581d7cc-298b-4c25-9dc2-40f23feb843b',
          rowNo: 9,
        },
      },
      {
        cursor: '63b84ad9fc13ae2b950000bd',
        node: {
          id: '63b84ad9fc13ae2b950000bd',
          type: 'debit',
          createdAt: '2021-11-11T01:04:40Z',
          amount: 66933.96,
          accountId: '78e39a2e-8568-46df-98aa-35d98d3fe76a',
          rowNo: 10,
        },
      },
    ],
    pageInfo: {
      endCursor: '63b84ad9fc13ae2b950000bd',
      hasNextPage: true,
    },
  },
};

let isError = false;
let isLoading = false;

jest.mock('../../src/hooks/useTransactions', () => ({
  __esModule: true,
  default: () => ({
    data: dummyTransactions,
    isLoading,
    isError,
  }),
}));

const setup = () => {
  const utils = render(
    <Transactions accountId="dummy_acount_id" pageSize={5} />
  );

  return {
    ...utils,
  };
};

describe('Transaction testing suite', () => {
  it('render without crashing', () => {
    const { asFragment, getByText, getByRole } = setup();

    expect(asFragment()).toMatchSnapshot();
    expect(getByText(/Page 1/i)).toBeInTheDocument();

    const pageHeading = getByText('List of transactions');
    expect(pageHeading).toBeInTheDocument();

    const previousBtn = getByRole('button', {
      name: 'Previous',
    });
    expect(previousBtn).toBeInTheDocument();
    expect(previousBtn).toBeDisabled();

    const nextBtn = getByRole('button', {
      name: 'Next',
    });
    expect(nextBtn).toBeInTheDocument();

    // click on next btn
    fireEvent.click(nextBtn);

    expect(getByText(/Page 2/i)).toBeInTheDocument();

    // click on prev btn
    fireEvent.click(previousBtn);

    expect(getByText(/Page 1/i)).toBeInTheDocument();
  });

  it('shows error alert in case of any error', () => {
    isError = true;
    const { getByText } = setup();
    expect(
      getByText(/Error while fetching transactions, please try again./i)
    ).toBeInTheDocument();
  });
});
