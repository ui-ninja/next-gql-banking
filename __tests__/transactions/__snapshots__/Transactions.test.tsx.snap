// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction testing suite render without crashing 1`] = `
<DocumentFragment>
  <div
    class="css-0"
  >
    <div
      class="chakra-alert css-1hwd9xd"
      role="alert"
    >
      <span
        class="chakra-alert__icon css-1c6vscd"
      >
        <svg
          class="chakra-icon css-qk6lof"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
            fill="currentColor"
          />
        </svg>
      </span>
      <div
        class="chakra-alert__title css-0"
      >
        This page uses dummy data!
      </div>
      <div
        class="chakra-alert__desc css-6n7j50"
      >
        Even though this page accepts accountId, it does not filter based on accountId due to lack of data points. But, this page follow cursor based pagination as per Relay specifications.
      </div>
    </div>
    <h1
      class="chakra-text css-1nvfbi4"
    >
      List of transactions
    </h1>
    <div
      class="chakra-table__container css-zipzvv"
    >
      <table
        class="chakra-table css-0"
      >
        <caption
          class="css-18i7bby"
        >
          Transactions data (5 per page)
        </caption>
        <thead
          class="css-0"
        >
          <tr
            class="css-0"
          >
            <th
              class="css-0"
            >
              S.no.
            </th>
            <th
              class="css-0"
            >
              ID
            </th>
            <th
              class="css-0"
            >
              Transaction done on
            </th>
            <th
              class="css-0"
            >
              Type
            </th>
            <th
              class="css-0"
            >
              Amount
            </th>
          </tr>
        </thead>
        <tbody
          class="css-0"
        >
          <tr
            class="css-0"
          >
            <td
              class="css-0"
            >
              1
            </td>
            <td
              class="css-0"
            >
              63b84ad9fc13ae2b950000b4
            </td>
            <td
              class="css-0"
            >
              2/21/2021, 5:35:34 AM
            </td>
            <td
              class="css-0"
            >
              credit
            </td>
            <td
              class="css-0"
            >
              80894.47
            </td>
          </tr>
          <tr
            class="css-0"
          >
            <td
              class="css-0"
            >
              2
            </td>
            <td
              class="css-0"
            >
              63b84ad9fc13ae2b950000b5
            </td>
            <td
              class="css-0"
            >
              4/21/2019, 2:39:46 PM
            </td>
            <td
              class="css-0"
            >
              credit
            </td>
            <td
              class="css-0"
            >
              39227.57
            </td>
          </tr>
          <tr
            class="css-0"
          >
            <td
              class="css-0"
            >
              3
            </td>
            <td
              class="css-0"
            >
              63b84ad9fc13ae2b950000b6
            </td>
            <td
              class="css-0"
            >
              2/6/2021, 6:34:12 PM
            </td>
            <td
              class="css-0"
            >
              credit
            </td>
            <td
              class="css-0"
            >
              86036.65
            </td>
          </tr>
          <tr
            class="css-0"
          >
            <td
              class="css-0"
            >
              4
            </td>
            <td
              class="css-0"
            >
              63b84ad9fc13ae2b950000b7
            </td>
            <td
              class="css-0"
            >
              2/15/2021, 10:47:32 AM
            </td>
            <td
              class="css-0"
            >
              credit
            </td>
            <td
              class="css-0"
            >
              7464.81
            </td>
          </tr>
          <tr
            class="css-0"
          >
            <td
              class="css-0"
            >
              5
            </td>
            <td
              class="css-0"
            >
              63b84ad9fc13ae2b950000b8
            </td>
            <td
              class="css-0"
            >
              9/6/2019, 11:50:27 AM
            </td>
            <td
              class="css-0"
            >
              credit
            </td>
            <td
              class="css-0"
            >
              7113.16
            </td>
          </tr>
          <tr
            class="css-0"
          >
            <td
              class="css-0"
            >
              6
            </td>
            <td
              class="css-0"
            >
              63b84ad9fc13ae2b950000b9
            </td>
            <td
              class="css-0"
            >
              10/9/2019, 10:42:04 PM
            </td>
            <td
              class="css-0"
            >
              credit
            </td>
            <td
              class="css-0"
            >
              12840.67
            </td>
          </tr>
          <tr
            class="css-0"
          >
            <td
              class="css-0"
            >
              7
            </td>
            <td
              class="css-0"
            >
              63b84ad9fc13ae2b950000ba
            </td>
            <td
              class="css-0"
            >
              8/24/2023, 7:49:50 AM
            </td>
            <td
              class="css-0"
            >
              debit
            </td>
            <td
              class="css-0"
            >
              18654.31
            </td>
          </tr>
          <tr
            class="css-0"
          >
            <td
              class="css-0"
            >
              8
            </td>
            <td
              class="css-0"
            >
              63b84ad9fc13ae2b950000bb
            </td>
            <td
              class="css-0"
            >
              7/18/2023, 12:01:06 AM
            </td>
            <td
              class="css-0"
            >
              debit
            </td>
            <td
              class="css-0"
            >
              14473.13
            </td>
          </tr>
          <tr
            class="css-0"
          >
            <td
              class="css-0"
            >
              9
            </td>
            <td
              class="css-0"
            >
              63b84ad9fc13ae2b950000bc
            </td>
            <td
              class="css-0"
            >
              1/24/2019, 5:50:04 PM
            </td>
            <td
              class="css-0"
            >
              debit
            </td>
            <td
              class="css-0"
            >
              96879.76
            </td>
          </tr>
          <tr
            class="css-0"
          >
            <td
              class="css-0"
            >
              10
            </td>
            <td
              class="css-0"
            >
              63b84ad9fc13ae2b950000bd
            </td>
            <td
              class="css-0"
            >
              11/11/2021, 6:34:40 AM
            </td>
            <td
              class="css-0"
            >
              debit
            </td>
            <td
              class="css-0"
            >
              66933.96
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="css-gmuwbf"
    >
      <button
        class="chakra-button css-4xx2wk"
        disabled=""
        type="button"
      >
        Previous
      </button>
      <h4
        class="chakra-text css-1k6rq5w"
      >
        Page 1
      </h4>
      <button
        class="chakra-button css-4xx2wk"
        type="button"
      >
        Next
      </button>
    </div>
  </div>
</DocumentFragment>
`;
